File: MECANUM_ROBOT_SDC-main/Reference/Robot_Swerve_SDC2025_AML-master/MATLAB/RobotAcceleration.m
classdef RobotAcceleration
    properties
        currentVelocity
        targetVelocity
        previousTarget
        duration
        tau
        timeElapsed
    end

    methods
        function obj = RobotAcceleration(initialTarget, duration)
            obj.currentVelocity = 0.0;
            obj.targetVelocity = initialTarget;
            obj.previousTarget = initialTarget;
            obj.duration = duration;
            obj.tau = duration / 5.0;
            obj.timeElapsed = 0.0;
        end

        function obj = setTarget(obj, newTargetVelocity, newDuration)
            obj.previousTarget = obj.currentVelocity;
            obj.targetVelocity = newTargetVelocity;
            obj.duration = newDuration;
            obj.tau = newDuration / 5.0;
            obj.timeElapsed = 0.0;
        end

        function [obj, velocity, acceleration] = update(obj, dt)
            obj.timeElapsed = obj.timeElapsed + dt;

            if obj.timeElapsed > obj.duration
                obj.timeElapsed = obj.duration;
            end

            velocityRange = obj.targetVelocity - obj.previousTarget;
            velocity = obj.previousTarget + velocityRange * (1 - exp(-obj.timeElapsed / obj.tau));
            acceleration = (velocityRange / obj.tau) * exp(-obj.timeElapsed / obj.tau);

            obj.currentVelocity = velocity; % Cập nhật vận tốc hiện tại
        end


    end

    methods (Static)
        function plotVelocityAndAcceleration(initialTarget, duration, dt)
            robot = RobotAcceleration(initialTarget, duration);
            time = 0:dt:duration;
            velocity = zeros(size(time));
            acceleration = zeros(size(time));

            for i = 1:length(time)
                [robot, velocity(i), acceleration(i)] = robot.update(dt);
            end

            figure;
            subplot(2,1,1);
            plot(time, velocity, 'b', 'LineWidth', 2);
            xlabel('Thời gian (s)');
            ylabel('Vận tốc (m/s)');
            title('Biểu đồ Vận tốc theo thời gian');
            grid on;

            subplot(2,1,2);
            plot(time, acceleration, 'r', 'LineWidth', 2);
            xlabel('Thời gian (s)');
            ylabel('Gia tốc (m/s^2)');
            title('Biểu đồ Gia tốc theo thời gian');
            grid on;
        end
    end
end

File: MECANUM_ROBOT_SDC-main/Core/Src/ROBOT_ACCELERATION.c
#include "ROBOT_ACCELERATION.h"
#include <math.h>

void RobotAcceleration_Init(RobotAcceleration *robot, float initialTarget, float duration)
{
  float safe_duration = (duration > 0.001f) ? duration : 0.001f;

  robot->currentVelocity = initialTarget;
  robot->targetVelocity = initialTarget;
  robot->previousTarget = initialTarget;
  robot->duration = safe_duration;
  robot->tau = safe_duration / 5.0f;
  robot->timeElapsed = safe_duration;
}

void RobotAcceleration_SetTarget(RobotAcceleration *robot, float newTargetVelocity, float newDuration)
{
  if (fabsf(newTargetVelocity - robot->targetVelocity) < 0.5f)
  {
    return;
  }

  robot->previousTarget = robot->currentVelocity;
  robot->targetVelocity = newTargetVelocity;

  {
    float safe_duration = (newDuration > 0.001f) ? newDuration : 0.001f;
    robot->duration = safe_duration;
    robot->tau = safe_duration / 5.0f;
  }

  robot->timeElapsed = 0.0f;
}

float RobotAcceleration_Update(RobotAcceleration *robot, float dt)
{
  robot->timeElapsed += dt;

  if (robot->timeElapsed >= robot->duration)
  {
    robot->timeElapsed = robot->duration;
    robot->currentVelocity = robot->targetVelocity;
    robot->previousTarget = robot->targetVelocity;
    return robot->currentVelocity;
  }

  if (robot->tau <= 0.00001f)
  {
    robot->currentVelocity = robot->targetVelocity;
    return robot->currentVelocity;
  }

  {
    float velocityRange = robot->targetVelocity - robot->previousTarget;
    robot->currentVelocity = robot->previousTarget + velocityRange * (1.0f - expf(-robot->timeElapsed / robot->tau));
  }

  return robot->currentVelocity;
}

